command/job_restart.go

```diff
@@ -0,0 +1,955 @@
+		default:
+		}
+
+		newAlloc, qm, err := c.client.Allocations().Info(nextAllocID, q)
+		if err != nil {
+			errCh <- fmt.Errorf("Failed to retrieve replacement allocation %q: %v", limit(nextAllocID, c.length), err)
+			return
+		}
+		if newAlloc.ClientStatus == api.AllocClientStatusRunning {
+			errCh <- nil
+			return
+		}
+		q.WaitIndex = qm.LastIndex
+	}
+}
+
+// handleSignal receives input signals and blocks the activeCh until the user
+// confirms how to proceed.
+//
+// Exit immediately if the user confirms the interrupt, otherwise resume the
+// command and feed activeCh to unblock it.
+func (c *JobRestartCommand) handleSignal(sigsCh chan os.Signal, activeCh chan any) {
+	for {
+		select {
+		case <-sigsCh:
+			// Consume activeCh to prevent the main loop from proceeding.
+			select {
+			case <-activeCh:
+			default:
+			}
+
+			if c.shouldExit() {
+				c.ui.Output("\nCanceling job restart process")
+				os.Exit(0)
+			}
+		case activeCh <- struct{}{}:
+		}
+	}
+}
+
+// shouldExit asks the user for confirmation if they would like to interrupt
+// the command. Returns true if the answer is positive.
+func (c *JobRestartCommand) shouldExit() bool {
+	if !isTty() {
+		return true
+	}
+
+	exitQuestion := `
+Are you sure you want to stop the restart process?
+Allocations not restarted yet will not be restarted. [y/N]`
+
+	for {
+		answer, err := c.ui.Ask(exitQuestion)
+		if err != nil {
+			if err.Error() != "interrupted" {
+				c.ui.Error(err.Error())
+			}
+			return true
+		}
+
+		switch strings.TrimSpace(strings.ToLower(answer)) {
+		case "y", "yes":
+			return true
+		case "n", "no", "":
+			return false
+		default:
+			c.ui.Output(fmt.Sprintf("Invalid option %q", answer))
+		}
+	}
+}
+
+// isErrNotRecoverable returns true when the error is likely to impact all
+// restarts and so there is not reason to keep going.
+func (c *JobRestartCommand) isErrNotRecoverable(err error) bool {
+	patterns := []string{
+		api.PermissionDeniedErrorContent,
+		jobRestartFailedToPlaceNewAllocation,
+	}
+	for _, pattern := range patterns {
+		if strings.Contains(err.Error(), pattern) {
+			return true
+		}
+	}
+	return false
+}
+
+// errorFormat is a multierror.ErrorFormatFunc that uses 2 spaces instead of a
+// tab before each error bullet point. This prevents deeply nested errors from
+// outputting very long lines.
+func (c *JobRestartCommand) errorFormat(es []error) string {
+	space := strings.Repeat(" ", 2)
+
+	if len(es) == 1 {
+		return fmt.Sprintf("1 error occurred:\n%s* %s\n\n", space, es[0])
+	}
+
+	points := make([]string, len(es))
+	for i, err := range es {
+		points[i] = fmt.Sprintf("* %s", err)
+	}
+
+	return fmt.Sprintf(
+		"%d errors occurred:\n%s%s\n\n",
+		len(es), space, strings.Join(points, fmt.Sprintf("\n%s", space)))
+}
+
+// formatSliceOf returns a string with the length and the given singular or
+// plural noun depending on how many elements are present.
+func formatSliceOf(in []string, singular string, plural string) string {
```

Comment:
Not in this PR but maybe for a follow-up, we have a bunch of these lovely formatting helpers all over the `command` package. It'd be nice if we could centralize them in a utils/helpers file so folks can find them and reuse them for UX consistency across commands.

Reply:
It turns out that I don't need this method anymore since we're not validating tasks and groups 😅 

But I will re-usability in mind for future commands.